### PR TITLE
Port 5831 Wiz integration powered by ocean

### PR DESCRIPTION
**Pull request by phalbert**
# Description

Add an integration to [Wiz](https://www.wiz.io/) that ingests the following objects:
- `Issue`: These are the main objects. Issues are the risks that Wiz has identified in the cloud environment, as defined by Controls, according to their severity.
- `Project`: Wiz Projects serve as a method to categorise resources based on users or their intended purposes. An `Issue` can affect multiple projects.
- `Control`: A security graph query defining a risk (e.g. "publicly exposed VM with effective global permissions") with a severity level (Critical, High, Medium, Low, or Info). Issues are generated by controls.
- `Service Ticket`: Automatically opened in the customer's project tracker e.g. Jira

## How Wiz works
Wiz detects basically everything running in your cloud infrastructure; configuration languages, frameworks, libraries, CI/CD tools, compute platforms, network, security, identity assets, databases, and almost every application component across workloads, accounts, and environments.

Wiz _continuously analyses_ configurations, vulnerabilities, networks, identities and access, secrets, and more across accounts, users, and workloads to **_discover the critical issues (`Issue`) that represent the actual risk._**

## Type of change

Please leave one option from the following and delete the rest:

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: integrations/wiz/client.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+            if not data:
+                logger.error(response_json.get("error"))
+                raise Exception("No data found for Wiz account")
+
+            issues = data["issues"]["nodes"]
+            event.attributes[CacheKeys.ISSUES] = issues
+            yield issues
+
+        except httpx.HTTPError as e:
+            logger.error(f"Error while fetching issues: {str(e)}")
```
Comment (yairsimantov20) [inline integrations/wiz/client.py:240]: why are u ignoring the exception? please raise it back

File: integrations/wiz/client.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+    def __init__(
+        self, api_url: str, client_id: str, client_secret: str, token_url: str
+    ):
+        self.api_url = api_url
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.token_url = token_url
+
+        self.http_client = http_async_client
+        self.http_client.headers.update(self.api_auth_header)
```
Comment (yairsimantov20) [inline integrations/wiz/client.py:142]: why are u setting the headers if u r applying it again every request

File: integrations/wiz/client.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+        audience = audience_mapping[self.token_url]
+
+        return {
+            "grant_type": "client_credentials",
+            "audience": audience,
+            "client_id": self.client_id,
+            "client_secret": self.client_secret,
+        }
+
+    def get_token(self) -> str:
```
Comment (yairsimantov20) [inline integrations/wiz/client.py:172]: take a look at how ocean holds port token

File: integrations/wiz/client.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+        try:
+            if cached_token_object := event.attributes.get(CacheKeys.TOKENS):
```
Comment (yairsimantov20) [inline integrations/wiz/client.py:174]: dont store the token in event cache
store it as lru cache

**Review by yairsimantov20 — CHANGES_REQUESTED** [review]
left some comments

File: integrations/wiz/client.py (hunk 1 above)
Comment (phalbert) [reply integrations/wiz/client.py:240]: I have raised it now

File: integrations/wiz/client.py (hunk 2 above)
Comment (phalbert) [reply integrations/wiz/client.py:142]: Thank you for catching this. I have removed it.

File: integrations/wiz/client.py (hunk 4 above)
Comment (phalbert) [reply integrations/wiz/client.py:174]: Do I still need to store it in a cache if I do it like ocean does?

File: integrations/wiz/client.py (hunk 3 above)
Comment (phalbert) [reply integrations/wiz/client.py:172]: I have considered it. Thank you.

File: integrations/wiz/client.py [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+            return TokenResponse(**response.json())
+        except Exception as e:
+            logger.exception(e)
+            raise
+
+    @property
+    async def token(self) -> str:
+        if not self.last_token_object or not self.last_token_object.is_valid:
+            msg = "Wiz Token expired or invalid, fetching new token"
+            logger.info(msg)
```
Comment (yairsimantov20) [inline integrations/wiz/client.py:204]:
```suggestion
            logger.info("Wiz Token expired or invalid, fetching new token")
```

File: integrations/wiz/client.py [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+            response = httpx.post(
+                self.token_url,
+                data=self.auth_params,
+                headers={"Content-Type": "application/x-www-form-urlencoded"},
+            )
+            response.raise_for_status()
+            return TokenResponse(**response.json())
+        except Exception as e:
+            logger.exception(e)
+            raise
```
Comment (yairsimantov20) [inline integrations/wiz/client.py:198]: no need
just let the exception get raised

File: integrations/wiz/client.py [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+
+    def _get_token(self) -> TokenResponse:
+        try:
+            response = httpx.post(
+                self.token_url,
+                data=self.auth_params,
+                headers={"Content-Type": "application/x-www-form-urlencoded"},
+            )
+            response.raise_for_status()
+            return TokenResponse(**response.json())
```
Comment (yairsimantov20) [inline integrations/wiz/client.py:195]: why isnt it using the client?

File: integrations/wiz/client.py [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+        self.api_url = api_url
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.token_url = token_url
+        self.last_token_object: TokenResponse | None = None
+
+        self.http_client = http_async_client
+
+    @property
+    async def api_auth_headers(self) -> dict[str, Any]:
```
Comment (yairsimantov20) [inline integrations/wiz/client.py:160]: please sort the functions in a reasonable way to read

File: integrations/wiz/client.py [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+
+    @property
+    def auth_params(self) -> dict[str, Any]:
+        audience_mapping = {
+            **{url: "beyond-api" for url in AUTH0_URLS},
+            **{url: "wiz-api" for url in COGNITO_URLS},
+        }
+
+        if self.token_url not in audience_mapping:
+            raise Exception("Invalid Token URL")
```
Comment (yairsimantov20) [inline integrations/wiz/client.py:176]: Dont use a broad exception

If there is a need for a custom exception then create a class that will inhirit ocean abort exception class and raise it with a proper exception so the client will understand how to fix it

File: integrations/wiz/client.py [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+            )
+            response.raise_for_status()
+            return TokenResponse(**response.json())
+        except Exception as e:
+            logger.exception(e)
+            raise
+
+    @property
+    async def token(self) -> str:
+        if not self.last_token_object or not self.last_token_object.is_valid:
```
Comment (yairsimantov20) [inline integrations/wiz/client.py:202]: what is `is_valid`? i think its a pydantic function
did you test this scenario?

File: integrations/wiz/client.py [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+
+            if not data:
+                logger.error(response_json.get("error"))
+                raise Exception("No data found for Wiz account")
+
+            issues = data["issues"]["nodes"]
+            event.attributes[CacheKeys.ISSUES] = issues
+            yield issues
+
+        except httpx.HTTPError as e:
```
Comment (yairsimantov20) [inline integrations/wiz/client.py:244]: more logs please

File: integrations/wiz/client.py [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+                json=json_data,
+                headers=await self.api_auth_headers,
+            )
+            response.raise_for_status()
+            response_json = response.json()
+            data = response_json.get("data")
+
+            if not data:
+                logger.error(response_json.get("error"))
+                raise Exception("No data found for Wiz account")
```
Comment (yairsimantov20) [inline integrations/wiz/client.py:238]: is there a case the status code will be ok and the response will have error?

File: integrations/wiz/client.py [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+            response = await self.http_client.post(
+                url=self.api_url,
+                json=json_data,
+                headers=await self.api_auth_headers,
+            )
+            response.raise_for_status()
+            response_json = response.json()
+            data = response_json.get("data")
+
+            if not data:
```
Comment (yairsimantov20) [inline integrations/wiz/client.py:236]: can the data be a list?

File: integrations/wiz/client.py [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+        if not self.last_token_object or not self.last_token_object.is_valid:
+            msg = "Wiz Token expired or invalid, fetching new token"
+            logger.info(msg)
+            self.last_token_object = await self._get_token(
+                self.client_id, self.client_secret
+            )
+
+        return self.last_token_object.token
+
+    async def get_issues(self) -> AsyncGenerator[list[dict[str, Any]], None]:
```
Comment (yairsimantov20) [inline integrations/wiz/client.py:211]: let me know what u think

We should either seperate the issues query from the projects or yield just unique projects because one project may have multiple issues which means that if i have 2000 issues then i will also try to resync the same project 2000 times

File: integrations/wiz/client.py [hunk 15]
```diff
@@ -0,0 +1,246 @@
...
+        logger.info("Fetching issues from Wiz API")
+
+        if cache := event.attributes.get(CacheKeys.ISSUES):
+            logger.info("picking Wiz issues from cache")
```
Comment (yairsimantov20) [inline integrations/wiz/client.py:215]:
```suggestion
            logger.info("Picking Wiz issues from cache")
```

File: integrations/wiz/main.py
Comment (yairsimantov20) [inline integrations/wiz/main.py:1]: no webhooks?

File: integrations/wiz/client.py (hunk 14 above)
Comment (yairsimantov20) [inline integrations/wiz/client.py:211]: where is the pagination

File: integrations/wiz/.port/spec.yaml [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+  - name: clientSecret
+    required: true
+    type: string
+    sensitive: true
+  - name: apiUrl
+    type: string
+    required: true
+  - name: tokenUrl
+    type: string
+    required: true
```
Comment (yairsimantov20) [inline integrations/wiz/.port/spec.yaml:23]: the convention so far is to prefix these kind of parameters with the integration name

File: integrations/wiz/.port/resources/port-app-config.yaml [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+      query: 'true'
+    port:
+      entity:
+        mappings:
+          blueprint: '"wizIssue"'
+          identifier: .id
+          title: .entitySnapshot.name + " | " + .entitySnapshot.type
+          properties:
+            # https://integrate.wiz.io/reference/issues-query#:~:text=Direct%20a%20user%20to%20an%20Issue%20in%20Wiz
+            url: .id as $id | "https://app.wiz.io/issues#~(issue~'" + $id + ")"
```
Comment (yairsimantov20) [inline integrations/wiz/.port/resources/port-app-config.yaml:28]: dont set things to a var if you not doing anything complex with them

```suggestion
            url: "https://app.wiz.io/issues#~(issue~'" + .id + ")"
```

File: integrations/wiz/.port/resources/port-app-config.yaml [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+            severity: .severity
+            type: .type
+            notes: .notes
+            vulnerability: .entitySnapshot
+            createdAt: .createdAt
+            updatedAt: .updatedAt
+            statusChangedAt: .statusChangedAt
+            resolvedAt: .resolvedAt
+          relations:
+            projects: .projects | map(.id)
```
Comment (yairsimantov20) [inline integrations/wiz/.port/resources/port-app-config.yaml:39]:
```suggestion
            projects: .project[].id'
```

File: integrations/wiz/.port/resources/port-app-config.yaml [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+            serviceTickets: .serviceTickets | (.externalId)
```
Comment (yairsimantov20) [inline integrations/wiz/.port/resources/port-app-config.yaml:40]: isnt it the same as?
```suggestion
            serviceTickets: .serviceTickets[].externalId
```

**Review by yairsimantov20 — CHANGES_REQUESTED** [review]
comments

File: integrations/wiz/client.py (hunk 12 above)
Comment (phalbert) [reply integrations/wiz/client.py:238]: There is a possibility of an unexpected internal server error happening and they return 200 with no data.


Error | Potential issues
-- | --
200 UNAUTHORIZED | The Wiz service account token scopes do not permit this action. Check the error message for the required scopes.
200 INTERNAL | An unexpected internal server error. Contact the Wiz Integrations team.

File: integrations/wiz/client.py (hunk 14 above)
Comment (phalbert) [reply integrations/wiz/client.py:211]: Oh! Yes, on a large scale, it becomes undesirable. I have asked @MPTG94 to help me get the `read project` scope added to the credentials I am using. We'll use the projects endpoint and separate them out.

File: integrations/wiz/.port/spec.yaml (hunk 16 above)
Comment (phalbert) [reply integrations/wiz/.port/spec.yaml:23]: alright then.

File: integrations/wiz/.port/resources/port-app-config.yaml (hunk 19 above)
Comment (phalbert) [reply integrations/wiz/.port/resources/port-app-config.yaml:40]: noted

File: integrations/wiz/.port/resources/port-app-config.yaml (hunk 18 above)
Comment (phalbert) [reply integrations/wiz/.port/resources/port-app-config.yaml:39]: noted here.

File: integrations/wiz/.port/resources/port-app-config.yaml (hunk 17 above)
Comment (phalbert) [reply integrations/wiz/.port/resources/port-app-config.yaml:28]: alright.

Comment (phalbert) [reply integrations/wiz/.port/resources/port-app-config.yaml:28]: I attempted to follow your suggestion, but the port configuration failed to register due to its invalidity. Ultimately, I found that what successfully registered the port-app-config in Port was the `var` approach.

File: integrations/wiz/client.py (hunk 5 above)
Comment (phalbert) [reply integrations/wiz/client.py:204]: noted this.

File: integrations/wiz/client.py (hunk 6 above)
Comment (phalbert) [reply integrations/wiz/client.py:198]: alright

File: integrations/wiz/client.py (hunk 8 above)
Comment (phalbert) [reply integrations/wiz/client.py:160]: I have now

File: integrations/wiz/client.py (hunk 9 above)
Comment (phalbert) [reply integrations/wiz/client.py:176]: alright, i have made this change

File: integrations/wiz/client.py (hunk 14 above)
Comment (phalbert) [reply integrations/wiz/client.py:211]: i have added it now

File: integrations/wiz/main.py
Comment (phalbert) [reply integrations/wiz/main.py:1]: now, we have the webhook. Although, there is no api for automatically creating the webhook at Wiz

File: integrations/wiz/.port/spec.yaml [hunk 20]
```diff
@@ -0,0 +1,27 @@
+type: wiz
+description: Wiz integration for Port Ocean
+icon: Alert
+features:
+  - type: exporter
+    section: Incident Management
+    resources:
+      - kind: issue
```
Comment (yairsimantov20) [inline integrations/wiz/.port/spec.yaml:8]: Wht about other resources?

**Review by yairsimantov20 — CHANGES_REQUESTED** [review]
Some more comments

File: integrations/wiz/client.py [hunk 21]
```diff
@@ -0,0 +1,352 @@
...
+from port_ocean.utils import http_async_client, get_time
+
+logger.configure(
+    handlers=[
+        {
+            "sink": sys.stdout,
+            "level": "INFO",
+            "format": "<red>Wiz Integration</red>: <green>[{time}]</green> | {level} | <level>{message}</level>",
+        }
+    ],
```
[the remark this answers is not in the document]
Comment (phalbert) [reply integrations/wiz/client.py:22]: I was debugging with this. I can removee it now.

File: integrations/wiz/client.py [hunk 22]
```diff
@@ -0,0 +1,352 @@
...
+      id
+      name
+      isFolder
+      archived
+      businessUnit
+      description
+    }
+  }
+}
+"""
```
[the remark this answers is not in the document]
Comment (phalbert) [reply integrations/wiz/client.py:164]: alright.

File: integrations/wiz/config.yaml [hunk 23]
```diff
@@ -0,0 +1,18 @@
+initializePortResources: true
+port:
+  clientId: "{{ from env PORT_CLIENT_ID }}"   # Can be loaded via environment variable: PORT_CLIENT_ID
+  clientSecret: "{{ from env PORT_CLIENT_SECRET }}"   # Can be loaded via environment variable: PORT_CLIENT_SECRET
+# The event listener to use for the integration service.
+eventListener:
+  type: POLLING
+integration:
+  # The identifier of this integration instance.
+  identifier: "{{ from env INTEGRATION_IDENTIFIER }}"
```
[the remark this answers is not in the document]
Comment (phalbert) [reply integrations/wiz/config.yaml:10]: I'll add a default: my-wiz-integration

File: integrations/wiz/.port/resources/port-app-config.yaml (hunk 17 above)
Comment (yairsimantov20) [reply integrations/wiz/.port/resources/port-app-config.yaml:28]: they syntax should be 
you had extra '
`"\"https://app.wiz.io/issues#~(issue~\" + .id + \")\""`

File: integrations/wiz/client.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (yairsimantov20) [reply integrations/wiz/client.py:22]: remove it pls

**Review by yairsimantov20 — CHANGES_REQUESTED** [review]

File: integrations/wiz/.port/spec.yaml (hunk 20 above)
Comment (phalbert) [reply integrations/wiz/.port/spec.yaml:8]: added them now

File: integrations/wiz/main.py [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+    return WizClient(
+        ocean.integration_config["wiz_api_url"],
+        ocean.integration_config["wiz_client_id"],
+        ocean.integration_config["wiz_client_secret"],
+        ocean.integration_config["wiz_token_url"],
+    )
+
+
+# initialise once
+wiz_client = init_client()
```
[the remark this answers is not in the document]
Comment (phalbert) [reply integrations/wiz/main.py]: Yeah, I have changed it now

File: integrations/wiz/client.py [hunk 25]
```diff
@@ -0,0 +1,352 @@
...
+        self, api_url: str, client_id: str, client_secret: str, token_url: str
+    ):
+        self.api_url = api_url
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.token_url = token_url
+        self.last_token_object: TokenResponse | None = None
+
+        self.http_client = http_async_client
+        self.http_client.timeout = Timeout(30)
```
[the remark this answers is not in the document]
Comment (phalbert) [reply integrations/wiz/client.py]: It was a test. I have removed it now

File: integrations/wiz/client.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (phalbert) [reply integrations/wiz/client.py]: I have removed it.

File: integrations/wiz/main.py [hunk 26]
```diff
@@ -0,0 +1,74 @@
...
+                    for ticket in issue.get("serviceTickets", [])
+                ]
+
+
+@ocean.router.post("/webhook")
+async def handle_webhook_request(
+    data: dict[str, Any], token: Any = Depends(HTTPBearer())
+) -> dict[str, Any]:
+    logger.info(f"Received token {token.__dict__}")
+    if ocean.integration_config.get("wiz_webhook_token") != token.credentials:
```
[the remark this answers is not in the document]
Comment (phalbert) [reply integrations/wiz/main.py]: I have corrected this now

File: integrations/wiz/.port/resources/port-app-config.yaml (hunk 17 above)
Comment (phalbert) [reply integrations/wiz/.port/resources/port-app-config.yaml:28]: the  extra ' exists in the documentation

Comment (phalbert) [reply integrations/wiz/.port/resources/port-app-config.yaml:28]: "\"https://app.wiz.io/issues#~(issue~\" + .id + \")\"" as well as other attempts I have made so far are failing to actually add the URL when i create the blueprint. This is why I kept the current way because it works alright. Lets stick with it.

**Comment by phalbert** [comment]
@yairsimantov20, The linting issue does not make sense. I see it also in a few other integrations like sonarqube, azure  integration

Screenshot 2024-01-17 at 11 37 22
<br />
<br />

I used codespaces to have a fresh dev env but still had that error on running `make run`


Screenshot 2024-01-17 at 11 38 39

File: integrations/wiz/client.py [hunk 27]
```diff
@@ -0,0 +1,205 @@
...
+
+class InvalidTokenUrlException(OceanAbortException):
+    def __init__(self, url: str, auth0_urls: list[str], cognito_urls: list[str]):
+        base_message = f"The token url {url} is not valid."
+        super().__init__(
+            f"{base_message} Valid token urls for AuthO are {auth0_urls} and for Cognito: {cognito_urls}"
+        )
+
+
+class UnexpectedWizException(BaseOceanException):
```
Comment (yairsimantov20) [inline integrations/wiz/client.py:35]: make it also inherit ocean abort

File: integrations/wiz/client.py [hunk 28]
```diff
@@ -0,0 +1,205 @@
...
+
+            response = await self.http_client.post(
+                url=self.api_url,
+                json={"query": query, "variables": variables},
+                headers=await self.auth_headers,
+            )
+            response_json = response.json()
+
+            if "data" not in response_json:
+                raise UnexpectedWizException(response_json["errors"])
```
Comment (yairsimantov20) [inline integrations/wiz/client.py:132]: wont raise for status handle this?

File: integrations/wiz/.port/spec.yaml [hunk 29]
```diff
@@ -0,0 +1,30 @@
...
+    required: true
+    type: string
+    sensitive: true
+  - name: wizApiUrl
+    type: string
+    required: true
+  - name: wizTokenUrl
+    type: string
+    required: true
+  - name: wizWebhookToken
```
Comment (yairsimantov20) [inline integrations/wiz/.port/spec.yaml:27]: rename here in the config.yaml and the main to wizWebhookVerificationToken

File: integrations/__init__.py
Comment (yairsimantov20) [inline integrations/__init__.py:1]: why did you add this file?

**Review by yairsimantov20 — CHANGES_REQUESTED** [review]
small comments

File: integrations/__init__.py
Comment (phalbert) [reply integrations/__init__.py:1]: Ah, this was the source of the linting problem all along. I think it was somehow added in one of those PyCharm auto improvements .

File: integrations/wiz/.port/spec.yaml (hunk 29 above)
Comment (phalbert) [reply integrations/wiz/.port/spec.yaml:27]: alright

**Review by yairsimantov20 — APPROVED** [review]
🌊
